biochatter/api_agent/blast.py

```diff
@@ -0,0 +1,352 @@
+
+        Args:
+            request_data: BlastQuery object containing the BLAST query
+                parameters.
+        Returns:
+            str: The Request ID (RID) for the submitted BLAST query.
+        """
+        data = {
+            "CMD": request_data.cmd,
+            "PROGRAM": request_data.program,
+            "DATABASE": request_data.database,
+            "QUERY": request_data.query,
+            "FORMAT_TYPE": request_data.format_type,
+            "MEGABLAST": request_data.megablast,
+            "HITLIST_SIZE": request_data.max_hits,
+        }
+        # Include any other_params if provided
+        if request_data.other_params:
+            data.update(request_data.other_params)
+        # Make the API call
+        query_string = urlencode(data)
+        # Combine base URL with the query string
+        full_url = f"{request_data.url}?{query_string}"
+        # Print the full URL
+        request_data.full_url = full_url
+        print("Full URL built by retriever:\n", request_data.full_url)
+        response = requests.post(request_data.url, data=data)
+        response.raise_for_status()
+        # Extract RID from response
+        print(response)
+        match = re.search(r"RID = (\w+)", response.text)
+        if match:
+            return match.group(1)
+        else:
+            raise ValueError("RID not found in BLAST submission response.")
+
+    def fetch_and_save_results(
+        self,
+        question_uuid: uuid,
+        query_return: str,
+        save_path: str,
+        max_attempts: int = 10000,
+    ):
+        """SECOND function to be called for a BLAST query.
+        Will look for the RID to fetch the data
+        """
+        file_name = f"BLAST_results_{question_uuid}.txt"
+        ###
+        ###    TO DO: Implement logging for all BLAST queries
+        ###
+        # log_question_uuid_json(request_data.question_uuid,question, file_name, save_path, log_file_path,request_data.full_url)
+        base_url = "https://blast.ncbi.nlm.nih.gov/Blast.cgi"
+        check_status_params = {
+            "CMD": "Get",
+            "FORMAT_OBJECT": "SearchInfo",
+            "RID": query_return,
+        }
+        get_results_params = {
+            "CMD": "Get",
+            "FORMAT_TYPE": "XML",
+            "RID": query_return,
+        }
+        # checking path: should end with '/'
+        if not save_path.endswith("/"):
+            save_path += "/"
+        # Check the status of the BLAST job
+        for attempt in range(max_attempts):
+            status_response = requests.get(base_url, params=check_status_params)
+            status_response.raise_for_status()
+            status_text = status_response.text
+            print("evaluating status")
+            if "Status=WAITING" in status_text:
+                print(f"{question_uuid} results not ready, waiting...")
+                time.sleep(15)
+            elif "Status=FAILED" in status_text:
```

Comment:
Do we need to quit loop if the query failed?

Reply:
If the query returns a FAILED status the elif is triggered and the loop is ended.

Reply:
I'm a little confused, in `elif "Status=FAILED" in status_text:` branch, we only write message to file without including `break` statement. How can we end the loop? Is this intentional?

Reply:
You are totally correct, I oversaw the missing break! thanks :)